readme.md
# Journal des décollages / treuillés de parapentistes  
Contributors: brouardt  
Author URI: https://thierry.brouard.pro/  
Plugin URL: https://thierry.brouard.pro/2018/01/paralog/  
Requires at Least: 3.1.0  
Tested Up To: 5.1.1  
Tags: treuil, ligne, parapente, pilote, journal  
Stable tag: 1.0  
License: GPLv2 or later  

Journal des décollages / treuillés.

## Description

Cette application vous permettra de tenir à jour, comme le recommande la FFVL, un journal des treuillés.

* gestion des sites.
* gestion des lignes ou lieux de décollage.
* gestion des pilotes.
* enregistrement et consultation des treuillés ou décollages.
* exportation des enregistrements au format csv.

L'application nécessite que l'utilisateur soit connecté au site, afin de l'identifier pour sauvegarder son identifiant lors des enregistrements de document et pour éviter que des systèmes automatiques viennent poluer le contenu. L'identification de l'utilisateur permettra à celui-ci de modifier ou de supprimer les documents dont il a été l'auteur. Les utilisateurs dont le rôles est éditeur peuvent modifier ou supprimer les documents de tous le monde. 
Les suppressions ne sont pas définitive. Les documents sont marqués comme supprimé. Seul l'administrateur pourra les retirer de la base.

### Developeurs

Cette extension a été développé initialement pour WordPress. Cependant rien ne vous empêche de reprendre le code et de l'adapter pour d'autres CMS.

Les développement se situent sur [GitHub](https://github.com/brouardt/paralog), alors n'hésitez pas à le reprendre.

## Installation
### Méthode 1 (via FTP)

1. Télécharger `paralog.zip`
2. Désarchivez `paralog.zip`
3. Téléverser le contenu de `paralog.zip` vers le dossier `/wp-content/plugins/`
4. Activez l'extension dans le menu 'Extensions' de WordPress.

### Méthode 2 (via ajout d'extension wordpress)

1. Télécharger `paralog.zip`
2. Téléverser `paralog.zip` dans la zone de dépots de fichier dans Extensions | ajouter | téléverser une extension | choisir un fichier
3. Cliquez sur installer
4. Activez l'extension dans le menu 'Extensions' de WordPress.

## Suivi des modifications

voir le fichier changelog.md


changelog.md
# Change Log
Contributors: brouardt  
Author URI: https://github.com/brouardt/paralog  

## [1.6.16] - 2019-06-15
## Correction
- checked au lieu de selected sur un radio

## [1.6.15] - 2019-06-15
## Modification
- Nettoyage + version

## [1.6.14] - 2019-06-15
## Modification
- Désactivation des limites d'alertes de présence pilote par date
- Ajout d'un état on/off de rappel de présence pilote.

## [1.6.13] - 2019-05-04
## Ajout
- Affichage des personnes non abonnés aux présences pilotes

## [1.6.12] - 2019-05-03
## Correction de bug
- Mauvais calcul du nombre de présence

## [1.6.11.1] - 2019-05-03
## Wording
- Modification du texte 'Once Weekly' par '1 fois par semaine'

## [1.6.11] - 2019-04-30
# Modification
- Tri de la vue des présences par date décroissante (défaut) 

## [1.6.10] - 2019-04-30
## Modification
- Arrangement de code

## [1.6.9] - 2019-04-30
## Ajout
- Ajout d'un bouton nouveau décollage

## [1.6.8] - 2019-04-30
## Correction
- Affichage de la vue des présences groupées par date

## [1.6.7] - 2019-04-30
## Correction
- Envoi e-mails multiple
## Modification
- Traduction

## [1.6.6] - 2019-04-29
## Ajout
- Prise en compte d'une date de rappel pour l'envoi manuel.

## [1.6.5] - 2019-04-26
## Ajout
- Gestion des appels de classes avec spl_autoloader_register()

## [1.6.1] - 2019-04-19
## Ajout
- Gestion des réglages
## Modification
- Élimination des shortcodes php
- Remplacement des die() en wp_die()

## [1.6.0] - 2019-04-15
## Ajout
- Gestion des abonnements de relance des pilotes, treuilleurs, élèves et moniteurs
## Modification
- Ajout des emails et bit d'abonnement aux rappels de présence

## [1.5.1] - 2019-04-14
## Ajout
- Création de la gestion de qui et quand sera présent
## Modification
- Protection des tables et champs

## [1.4.9] - 2019-03-28
## Ajout
- Création de la gestion des activités
- Nouvelle structure de base de données

## [1.3.11] - 2018-09-02
## Modification
- Modification de la gestion des définitions de cookies

## [1.3.10] - 2018-08-25
## Ajout
- Saisie de la date et de l'heure de décollage / treuillage désormais possible.

## [1.3.9] - 2018-04-28
## Correction
- Correction d'un bug du calcul statistique du nombre de passagers sur sites.

## [1.3.8] - 2018-04-28
## Correction
- Ajout du filtre deleted = 0 dans les requêtes nécessitant sont usage.

## [1.3.7] - 2018-04-27
## Modification
- Gestion des suppressions douces. Celles-ci seront seulement dans un état deleted avant 
que l'administrateur ne les supprime définitivement

## [1.3.6] - 2018-04-22
## Modfication
- Prise en compte des auteurs de documents (site, ligne, personnel, décollage). Ceux-ci pourront désormais modifier ou supprimer leurs documents.

## [1.3.5] - 2018-03-22
## Modification
- Prise en compte du format réel de la date et heure définie dans wordpress
## Suppression
- Retrait des traductions du format de date et heure

## [1.3.4] - 2018-03-15
## Modification
- Retrait de la colonne user_id dans l'exportation des données

## [1.3.3] - 2018-03-15
## Modification
- Changements dans les textes et traductions.

## [1.3.2] - 2018-03-12
## Ajout
- Dans la vue Personne, le numéro de licence est lié au site de la FFVL.

## [1.3.1] - 2018-03-11
## Ajout
- Insertion du user_id actif dans les tables de sauvegarde.

## [1.2.9] - 2018-03-11
### Ajout
- Création d'un cookie pour memorisation temporaire (durée: 12H) du site et de la ligne de vol. Afin d'éviter toute saisies superflues dans le journal de log, lorsque l'on se trouve toute la journée sur le même site.

## [1.2.8] - 2018-03-06
### Modification
- Exportation des données par dates de décollage décroissantes

## [1.2.7] - 2018-03-04
### Modification
- Amélioration de l'export. Extraction des noms de colonnes de table automatique et exportation sequentielle.

## [1.2.4] - 2018-03-03
### Ajout
- Exportation des données brutes depuis la page "A propos de"

## [1.2.3] - 2018-03-01
### Modification
- La première colonne des personnes regroupe désormais le prénom et le nom. Ce qui rend plus lisible les pilotes sur écran de téléphone
- Suppression de la visualisation de la colonne n° vol

## [1.2.1] - 2018-02-25
### Modification
- La première colonne des journaux de décollage devient la date et heure de décollage

## [1.1.7] - 2018-02-19
### Correction
- La gestion de la sauvegarde du champ "passenger_name" type VARCHAR(129) DEFAULT NULL ne prenait pas correctement en compte la valeur (vide).

## [1.1] - 2018-01-02
### Ajout
- formulaire (ajout et modification) de line, personne et journaux de treuillé.

## [1.0.135] - 2018-01-01
### Ajout
- formulaire (ajout et modification) de site

## [1.0.108] - 2017-12-30
### Ajout
- admin menu (A propos de)
- statistiques des éléments des tables wl-...
- visualisation des éléments sites, lignes, personnes, treuillés

## [1.0.60] - 2017-12-01
## Ajout
- L'extention peut être installée, désactivée, désinstallée
- Écriture des paramètres : db_version, active/inactive, version

## [1.0.0] - 2017-11-30
### Ajout
- Création du squelette de l'application en mode Objet.


includes/paralog_options.php
<?php
if ( ! defined( 'ABSPATH' ) ) {
	wp_die( 'No direct access allowed', 'Security' );
}

/**
 * @package Paralog_Options
 *
 * @author Thierry Brouard <thierry@brouard.pro>
 */
class Paralog_Options
{
	const RAISE = 'paralog_raise';
	const RECURRENCE = 'weekly';


	/**
	 * @param $schedules
	 * @return array
	 */
	public static function add_weekly_to_cron( $schedules )
	{
		if ( ! isset( $schedules[ self::RECURRENCE ] ) ) {
			$schedules[ self::RECURRENCE ] = array(
				'interval' => 3600 * 24 * 7, // 3600s = 1h x 24h x 7j
				'display' => __( 'Une fois par semaine', PL_DOMAIN )
			);
		}

		return $schedules;
	}

	public function form_edit()
	{
		$message = '';
		$notice = '';

		$default = array(
			'activity' => 1,
			'raise_day' => 'Friday',
			'raise_time' => '12:00 UTC',
			'raise_subject' => '',
			'raise_message' => '',
		);

		if ( isset( $_REQUEST['nonce'] ) && wp_verify_nonce( $_REQUEST['nonce'], basename( __FILE__ ) ) ) {
			// combine our default item with request params
			$request = stripslashes_deep( $_REQUEST );
			$item = shortcode_atts( $default, $request );

			$item_valid = $this->form_validate( $item );

			if ( $item_valid === true ) {
				if ( $_REQUEST['submit'] === 'save' ) {
					$options = get_option( PL_DOMAIN );
					foreach ( array_keys( $default ) as $key ) {
						$options[ $key ] = $item[ $key ];
					}
					update_option( PL_DOMAIN, $options, 'no' );
					self::remove_cron();
					self::add_cron();
					$message = __( 'Paramètres mis à jour', PL_DOMAIN );
				}
				if ( $_REQUEST['submit'] === 'send' ) {
					$date = filter_input( INPUT_POST, 'raise_date' );
					$count = self::raise_pilots( $date );
					if ( $count ) {
						$message = "$count " . __( 'rappel(s) envoyé(s)', PL_DOMAIN );
					} else {
						$notice = __( "Un problème empeche l'envoi de l'e-mail", PL_DOMAIN );
					}
				}
			} else {
				// if $item_valid not true it contains error message(s)
				$notice = $item_valid;
			}
		} else {
			// if this is not post back we load item to edit or give new one to create
			$item = $default;
			$options = get_option( PL_DOMAIN );
			foreach ( array_keys( $default ) as $key ) {
				$item[ $key ] = $options[ $key ];
			}
		}
		add_meta_box( 'options_form_meta_box', __( 'Paramètre', PL_DOMAIN ), array(
			$this,
			'options_form_meta_box_handler',
		), 'options', 'normal', 'default' );
		$date_raise = new DateTime( 'tomorrow' );
		$date_raise_min = $date_raise->format( 'Y-m-d' );
		?>
        <div class="wrap">
            <div class="icon32 icon32-posts-post" id="icon-edit"></div>
            <h1><?php _e( "Réglages", PL_DOMAIN ); ?></h1>
            <div class="notice notice-info is-dismissible">
				<?php _e( "Vous pouvez utiliser des variables pour un traitement dynamique de l'information dans le sujet ou le message de votre e-mail.", PL_DOMAIN ); ?>
                <ul class="ul-square">
                    <li>%DATE_ATTENDANCE% = <?php _e( "la date du lendemain.", PL_DOMAIN ); ?></li>
                    <li>%FORM_ATTENDANCE%
                        = <?php _e( "lien vers le formulaire de saisie pré-rempli.", PL_DOMAIN ); ?></li>
                    <li>%UNSUBSCRIBE_ATTENDANCE%
                        = <?php _e( "lien vers la fiche pilote pour modification de paramètres.", PL_DOMAIN ); ?></li>
                </ul>
            </div>
			<?php if ( ! empty( $notice ) ): ?>
                <div id="notice" class="error">
                    <p>
						<?php echo $notice; ?>
                    </p>
                </div>
			<?php endif; ?>
			<?php if ( ! empty( $message ) ): ?>
                <div id="message" class="updated">
                    <p>
						<?php echo $message; ?>
                    </p>
                </div>
			<?php endif; ?>
            <form id="form" method="post">
                <input type="hidden" name="nonce" value="<?php echo wp_create_nonce( basename( __FILE__ ) ); ?>"/>
                <div class="metabox-holder" id="postoptions">
                    <div id="post-body">
                        <div id="post-body-content">
							<?php do_meta_boxes( "options", 'normal', $item ); ?>
                            <button type="submit" name="submit" value="save" class="button button-primary">
                                <span class="fa fa-save"></span>
								<?php _e( 'Enregistrer', PL_DOMAIN ); ?>
                            </button>
                            <button type="submit" name="submit" value="send" class="button button-secondary">
                                <span class="fa fa-send"></span>
								<?php _e( 'Envoyer un rappel pour le ', PL_DOMAIN ); ?>
                            </button>
                            <input type="date" name="raise_date" value="<?php echo esc_attr( $date_raise_min ); ?>"
                                   min="<?php echo esc_attr( $date_raise_min ); ?>" />
                        </div>
                    </div>
                </div>
            </form>
        </div>
		<?php
	}

	private function form_validate( $item )
	{
		$messages = array();

		if( $item['activity'] == '' ) {
            $messages[] = __( "Le rappel est à définir", PL_DOMAIN );
        }

		if ( empty( $messages ) ) {
			return true;
		}

		return implode( '<br />', $messages );
	}

	public static function remove_cron()
	{
		$timestamp = wp_next_scheduled( self::RAISE );
		wp_unschedule_event( $timestamp, self::RAISE );
	}

	public static function add_cron()
	{
		$options = get_option( PL_DOMAIN );

		if ( isset( $options['raise_day'] ) && $options['raise_day'] != '' ) {
			if ( ! wp_next_scheduled( self::RAISE ) ) {

				$raise = 'next ' . $options['raise_day'] .
				         ( isset( $options['raise_time'] ) ? $options['raise_time'] : '12:00 UTC' );
				$timestamp = strtotime( $raise );
				wp_schedule_event( $timestamp, self::RECURRENCE, self::RAISE );
			}
			add_action( self::RAISE, array( __CLASS__, 'raise_pilots' ) );
		}
	}

	public static function raise_pilots( $date = 'tomorrow' )
	{
		global $wpdb;

		$options = get_option( PL_DOMAIN );

		if ( $options['activity'] == 1 ) {
			$table = Paralog::table_name( 'persons' );
			$query = "SELECT " .
			         "`person_id`, " .
			         "`firstname`, " .
			         "`lastname`, " .
			         "`email` " .
			         "FROM `$table` " .
			         "WHERE `deleted` = 0 " .
			         "AND `raise` = 1";
			$recipients = $wpdb->get_results( $query );

			$date = new DateTime( $date, new DateTimeZone( get_option( 'timezone_string' ) ) );
			$date_attendance = $date->format( get_option( 'date_format' ) );

			$count = 0;
			foreach ( $recipients as $recipient ) {
				$to = ltrim( "{$recipient->firstname} {$recipient->lastname} <{$recipient->email}>" );

				$form_attendance = get_admin_url(
					get_current_blog_id(),
					sprintf(
						'admin.php?page=paralog-attendances-form&date=%s&person_id=%d',
						$date->format( 'Y-m-d' ),
						$recipient->person_id
					)
				);

				$unsubscribe_attendance = get_admin_url(
					get_current_blog_id(),
					sprintf(
						'admin.php?page=paralog-persons-form&id=%d',
						$recipient->person_id
					)
				);

				$subject = str_ireplace(
					array( '%DATE_ATTENDANCE%', '%FORM_ATTENDANCE%', '%UNSUBSCRIBE_ATTENDANCE%' ),
					array( $date_attendance, $form_attendance, $unsubscribe_attendance ),
					$options['raise_subject']
				);

				$message = str_ireplace(
					array( '%DATE_ATTENDANCE%', '%FORM_ATTENDANCE%', '%UNSUBSCRIBE_ATTENDANCE%' ),
					array( $date_attendance, $form_attendance, $unsubscribe_attendance ),
					$options['raise_message']
				);
				/*
				 * envoi du mail
				 */
				wp_mail( $to, $subject, $message );
				$count ++;
			}

			return $count;
		}

		return false;
	}

	public function options_form_meta_box_handler( $item )
	{
		$days = array(
			'' => __( 'Aucun', PL_DOMAIN ),
			'Monday' => __( 'Lundi', PL_DOMAIN ),
			'Tuesday' => __( 'Mardi', PL_DOMAIN ),
			'Wednesday' => __( 'Mercredi', PL_DOMAIN ),
			'Thursday' => __( 'Jeudi', PL_DOMAIN ),
			'Friday' => __( 'Vendredi', PL_DOMAIN ),
			'Saturday' => __( 'Samedi', PL_DOMAIN ),
			'Sunday' => __( 'Dimanche', PL_DOMAIN ),
		);
		$times = array(
			'00:00 UTC' => __( "00:00 UTC", PL_DOMAIN ),
			'01:00 UTC' => __( "01:00 UTC", PL_DOMAIN ),
			'02:00 UTC' => __( "02:00 UTC", PL_DOMAIN ),
			'03:00 UTC' => __( "03:00 UTC", PL_DOMAIN ),
			'04:00 UTC' => __( "04:00 UTC", PL_DOMAIN ),
			'05:00 UTC' => __( "05:00 UTC", PL_DOMAIN ),
			'06:00 UTC' => __( "06:00 UTC", PL_DOMAIN ),
			'07:00 UTC' => __( "07:00 UTC", PL_DOMAIN ),
			'08:00 UTC' => __( "08:00 UTC", PL_DOMAIN ),
			'09:00 UTC' => __( "09:00 UTC", PL_DOMAIN ),
			'10:00 UTC' => __( "10:00 UTC", PL_DOMAIN ),
			'11:00 UTC' => __( "11:00 UTC", PL_DOMAIN ),
			'12:00 UTC' => __( "12:00 UTC", PL_DOMAIN ),
			'13:00 UTC' => __( "13:00 UTC", PL_DOMAIN ),
			'14:00 UTC' => __( "14:00 UTC", PL_DOMAIN ),
			'15:00 UTC' => __( "15:00 UTC", PL_DOMAIN ),
			'16:00 UTC' => __( "16:00 UTC", PL_DOMAIN ),
			'17:00 UTC' => __( "17:00 UTC", PL_DOMAIN ),
			'18:00 UTC' => __( "18:00 UTC", PL_DOMAIN ),
			'19:00 UTC' => __( "19:00 UTC", PL_DOMAIN ),
			'20:00 UTC' => __( "20:00 UTC", PL_DOMAIN ),
			'21:00 UTC' => __( "21:00 UTC", PL_DOMAIN ),
			'22:00 UTC' => __( "22:00 UTC", PL_DOMAIN ),
			'23:00 UTC' => __( "23:00 UTC", PL_DOMAIN )
		);
		?>
        <table cellspacing="2" cellpadding="5" style="width: 100%;" class="form-table">
            <tbody>
            <tr>
                <th valign="top" scope="row">
                    <label>
			            <?php _e( "Rappel de l'activité", PL_DOMAIN ); ?>
                    </label>
                </th>
                <td>
                    <label>
                        <input name="activity" type="radio"
                               value="1" <?php echo ($item['activity'] == 1 ? 'checked': ''); ?> class="code" />
                        <?php _e('Oui', PL_DOMAIN); ?>
                    </label>
                    <label>
                        <input name="activity" type="radio"
                               value="0" <?php echo ($item['activity'] == 0 ? 'checked': ''); ?> class="code" />
                        <?php _e('Non', PL_DOMAIN); ?>
                    </label>
                </td>
            </tr>
            <tr class="form-field">
                <th valign="top" scope="row">
                    <label for="raise_day">
						<?php _e( "Jour du rappel", PL_DOMAIN ); ?>
                    </label>
                </th>
                <td>
                    <select id="raise_day" name="raise_day">
						<?php foreach ( $days as $key => $value ): ?>
                            <option value="<?php echo esc_attr( $key ); ?>"
								<?php echo ( $key == $item['raise_day'] ? 'selected' : '' ); ?>>
								<?php echo esc_html( $value ); ?>
                            </option>
						<?php endforeach; ?>
                    </select>
                </td>
            </tr>
            <tr class="form-field">
                <th valign="top" scope="row">
                    <label for="raise_time">
						<?php _e( "Heure du rappel", PL_DOMAIN ) ?>
                    </label>
                </th>
                <td>
                    <select id="raise_time" name="raise_time">
						<?php foreach ( $times as $key => $value ): ?>
                            <option value="<?php echo esc_attr( $key ); ?>"
								<?php echo ( $key == $item['raise_time'] ? 'selected' : '' ); ?>>
								<?php echo esc_html( $value ); ?>
                            </option>
						<?php endforeach; ?>
                    </select>
                </td>
            </tr>
            <tr class="form-field">
                <th valign="top" scope="row">
                    <label for="raise_subject">
						<?php _e( 'Sujet de rappel', PL_DOMAIN ); ?>
                    </label>
                </th>
                <td>
                    <input type="text" name="raise_subject"
                           value="<?php echo esc_html( $item['raise_subject'] ); ?>" class="code"/>
                </td>
            </tr>
            <tr class="form-field">
                <th valign="top" scope="row">
                    <label for="raise_message">
						<?php _e( 'Message de rappel', PL_DOMAIN ); ?>
                    </label>
                </th>
                <td>
                    <textarea name="raise_message" class="code"
                              rows="5"><?php echo esc_html( $item['raise_message'] ); ?></textarea>
                </td>
            </tr>
            </tbody>
        </table>
		<?php
	}
}
